cmd/soroban-cli/src/commands/contract/deploy/wasm.rs

```diff
@@ -100,19 +108,91 @@ pub enum Error {
     Network(#[from] network::Error),
     #[error(transparent)]
     Wasm(#[from] wasm::Error),
+    #[error("cannot access config dir for alias file")]
+    CannotAccessConfigDir,
+    #[error(
+        "alias must be 1-30 chars long, and have only letters, numbers, underscores and dashes"
+    )]
+    InvalidAliasFormat { alias: String },
+    #[error(transparent)]
+    JSONSerialization(#[from] serde_json::Error),
+    #[error(transparent)]
+    IO(#[from] std::io::Error),
 }
 
 impl Cmd {
     pub async fn run(&self) -> Result<(), Error> {
+        self.validate_alias()?;
+
         let res = self.run_against_rpc_server(None, None).await?.to_envelope();
         match res {
             TxnEnvelopeResult::TxnEnvelope(tx) => println!("{}", tx.to_xdr_base64(Limits::none())?),
             TxnEnvelopeResult::Res(contract) => {
+                self.save_contract_id(&contract)?;
                 println!("{contract}");
             }
         }
         Ok(())
     }
+
+    fn validate_alias(&self) -> Result<(), Error> {
+        let alias = self.alias();
+
+        if alias.is_empty() {
+            return Ok(());
+        }
+
+        let regex = Regex::new(r"^[a-zA-Z0-9_-]{1,30}$").unwrap();
+
+        if regex.is_match(&alias) {
+            Ok(())
+        } else {
+            Err(Error::InvalidAliasFormat { alias })
+        }
+    }
+
+    fn alias(&self) -> String {
+        self.alias.as_ref().map(Clone::clone).unwrap_or_default()
+    }
+
+    fn alias_path(&self) -> Result<PathBuf, Error> {
+        let config_dir = self.config.config_dir()?;
+        let network = self.config.network.network.clone().expect("must be set");
+        let alias = self.alias();
+        let file_name = format!("{alias}.json");
+
+        Ok(config_dir
+            .join("contract-ids")
+            .join(network)
+            .join(file_name))
+    }
+
+    fn save_contract_id(&self, contract: &String) -> Result<(), Error> {
+        let file_path = self.alias_path()?;
+
+        let Some(dir) = file_path.parent() else {
+            return Err(Error::CannotAccessConfigDir);
+        };
+
+        match create_dir_all(dir) {
+            Ok(()) => {}
+            _ => return Err(Error::CannotAccessConfigDir),
+        };
+
+        let mut to_file = OpenOptions::new()
+            .create(true)
+            .write(true)
+            .open(file_path)
+            .map_err(Error::IO)?;
+
+        let payload = AliasData {
+            id: contract.into(),
+        };
+
+        let content = serde_json::to_string(&payload).map_err(Error::JSONSerialization)?;
+
+        to_file.write_all(content.as_bytes()).map_err(Error::IO)
```

Comment:
Something interesting is happening here that initially wasn't obvious to me.

When no alias is set, the code still runs to write it to a file with an empty name. This happens because the alias is unwrap_or_default resulting in an empty string. Some locations that then use that value are guarded from the case it is empty, but not all, leading to the file writing code running.

I think should remove the unwrap_or_default, and instead use the Option<String> in all the places.

Reply:
huh. I couldn't reproduce that. What file name is being used, given that `alias` is not set? I'd expect `.soroban/contract-ids/testnet/.json`, but again, couldn't reproduce it.

Reply:
I swear I had a check in place. It probably got removed during a refactor. Fixed!